readme.md
# otterspace-contracts

[![unit tests](https://github.com/otterspace-xyz/otterspace-contracts/actions/workflows/main.yml/badge.svg)](https://github.com/otterspace-xyz/otterspace-contracts/actions/workflows/main.yml)

### Otterspace’s [EIP-4973](https://github.com/ethereum/EIPs/pull/4973)-compliant non-transferable badge protocol helps DAOs create better incentive systems, automate permissions and enable non-financialized governance ✨ 🦦 🚀

---

# Project setup

- Install [Foundry](https://book.getfoundry.sh/getting-started/installation.html)
- Requires Node `v16` and Solidity `0.8.16`

```bash
git clone git@github.com:otterspace-xyz/otterspace-contracts.git
git submodule update --init
yarn
forge install
forge build
forge test
npx hardhat typechain
npx hardhat test
```

### Supported networks and contract addresses
Checkout our documentation site at [docs.otterspace.xyz](https://docs.otterspace.xyz/documentation/contracts)

### Use ABIs with JavaScript

We're publishing this repository at [@otterspace-xyz/contracts](https://www.npmjs.com/package/@otterspace-xyz/contracts)

```bash
npm i @otterspace-xyz/contracts
```

With node >= 16, contract ABIs can be imported into JavaScript applications as npm dependencies as follows:

Badges ABI

```js
import Badges from '@otterspace-xyz/contracts/out/Badges.sol/Badges.json' assert { type: 'json' }
```

Raft ABI

```js
import Raft from '@otterspace-xyz/contracts/out/Raft.sol/Raft.json' assert { type: 'json' }
```

SpecDataHolder ABI

```js
import SpecDataHolder from '@otterspace-xyz/contracts/out/SpecDataHolder.sol/SpecDataHolder.json' assert { type: 'json' }
```

We're exporting specific `.sol` files using the `"files"` property in
`package.json`. Please familiarize yourself with the `.sol` files we're
exporting by looking into `package.json`.

---

## Tech Stack

We use **Foundry** and **Hardhat** together. With this setup we get:

- Unit tests written in Solidity (Forge)
- Integration tests written in JavaScript (Mocha)

---

## Foundry

**Foundry is a blazing fast, portable and modular toolkit for Ethereum application development written in Rust.**

Foundry consists of:

- [**Forge**](https://book.getfoundry.sh/forge/): Ethereum testing framework (like Truffle, Hardhat and DappTools).
- [**Cast**](https://book.getfoundry.sh/cast/): Swiss army knife for interacting with EVM smart contracts, sending transactions and getting chain data.
- [**Anvil**](https://book.getfoundry.sh/anvil/): local Ethereum node, akin to Ganache, Hardhat Network.

**Need help getting started with Foundry? Read the [📖 Foundry Book](https://book.getfoundry.sh/)!**

### Hardhat

Hardhat is an Ethereum development environment for professionals. It facilitates performing frequent tasks, such as running tests, automatically checking code for mistakes or interacting with a smart contract.

On [Hardhat's website](https://hardhat.org) you will find:

- [Guides to get started](https://hardhat.org/getting-started/)
- [Hardhat Network](https://hardhat.org/hardhat-network/)
- [Plugin list](https://hardhat.org/plugins/)

## Deploying contracts for the first time

- create a `.env` file matching the variables seen in `.env.example`
- run `./scripts/deployProxy.ts .env`
- Hardhat will deploy the SpecDataHolder, Raft, and Badges contracts, then deploy a proxy for each one.
- Once deployed, follow the logged instructions in your terminal to verify the contracts.
- it will tell you: `npx hardhat verify --network ${networkName} ${contractAddress}`

## Testing upgradability

- Our contracts use the OpenZeppelin [UUPS](https://docs.openzeppelin.com/contracts/4.x/api/proxy#UUPSUpgradeable) proxy pattern for upgrades.
- To test your working changes against the latest release:
  1. make sure you're on the latest from the `dev` branch
  2. run `forge build` to make sure you've generated the latest ABIs
  3. copy the contents of:
    - `artifacts/src/Badges.sol/Badges.json` into `src/test/abis/latest`
    - `artifacts/src/Raft.sol/Raft.json` into `src/test/abis/latest`
    - `artifacts/src/SpecDataHolder.sol/SpecDataHolder.json` into `src/test/abis/latest`
- 4. make some changes to the contracts
- 5. run `npx hardhat testUpgrade`
- 6. As long as you don't see errors, your new contract is upgrade safe!

## Running upgrades (overview)

- Once you've confirmed that the contracts are upgrade safe you'll
  - deploy an implementation (not proxy) of each changed contract
  - point the proxy at the new implementation

## Running upgrades (steps)

- make sure your `.env.implementation` file has the correct values
- run `./scripts deploy_and_verify_implementation.sh .env.implementation ${contractName} ${networkName}`
- to deploy Badges on Optimism it would be
  - `./scripts deploy_and_verify_implementation.sh .env.implementation Badges optimism`
- watch the console for confirmation of deployment and verification, copy the implementation's address
- go into Defender and propose and execute the upgrade

## How Otterspace runs upgrades
- Our process of running upgrades is managed by github actions, found in the `.github/workflows` directory.

## Publishing attestations

`npx hardhat run publishAttestations.js --network optimisticEthereum`

more details about in `./scripts/update_scores.py`

## Foundry setup for VS Code Users

Add a `.vscode` file under the root

```json
{
  "solidity.packageDefaultDependenciesContractsDirectory": "src",
  "solidity.packageDefaultDependenciesDirectory": "lib",
  "solidity.compileUsingRemoteVersion": "v0.8.16"
}
```

## Versioning & publishing to NPM
- First update the version in package.json
- Add the version entry in changelog.md
- Commit all your changes
- Run `git tag v5.x.x` to the version you added in package.json
- Run `git push origin v5.x.x`
- Run `npm publish`

Note - you need to have admin rights to otterspace registry in npm

## Changelog

See changelog.md file.

## Supporting new chains
- [Adding support for new chains](https://www.notion.so/otterspace-xyz/How-to-add-support-for-a-new-chain-7f40b88185304a248f8014a7fd0af6c4)


## License

See LICENSE file.


changelog.md
# Changelog

We're using https://semver.org/

## 2.8.0
- Introduces functionality to update Badge's token uri and emits a `MetadataUpdate` event

## 2.7.3
- Bug fix in `mintWithConsent()` and rebuilding faulty ABIs

## 2.7.2

- add `merkleMintWithConsent()`

## 2.7.1

- bug fix in agreementHash in `mintWithConsent()`

## 2.7.0

- update `mintWithConsent()` to not use msg.sender

## 2.6.7

- Add `mintWithConsent()`

## 2.6.6

- Change `addAdmins` and `removeAdmins` on Raft contract to `setAdmins`

## 2.6.5

- Add `addAdmins` and `removeAdmins` to Raft contract

## 2.6.4

- Add `setBadgesToRafts` and `setSpecsToRafts`

## 2.6.3

- Add `giveRequestedBadgeToMany` and tests

## 2.6.2

- Allow Raft admins to airdrop badges

## 2.6.1

- All minting actions now rely on checking admin list for authorization

## 2.6.0

- Raft contract now supports adding multiple admins via `isAdminActive` and `setAdmin` functions.
- A new minting functionality is added with `giveRequestedBadge` where any of the admins can issue a requested badge
- giveRequestBadge admin authorization by checking with `isAdminActive`

## 2.5.1

- Add modifier to setSpecToRaft

## 2.5.0

- Add multi-voucher issuance

## 2.4.0

- Update the "give" function to make sure that the caller is the owner of the Raft token associated with the targeted badge spec

## 2.3.0

- Add a function that enables us refresh a badge's metadata by emitting an event

## 2.2.0

- Add revocation and reinstatement of badges

## 2.1.0

- Added MetadataUpdate event in Raft.sol when setTokenURI is called

## 2.0.1

- Update Solidity version in Raft.sol

- Live contracts on **Optimism**
- - [Badges.sol](https://optimistic.etherscan.io/address/0x7F9279B24D1c36Fa3E517041fdb4E8788dc63D25)
- - [Raft.sol](https://optimistic.etherscan.io/address/0xa6773847d3D2c8012C9cF62818b320eE278Ff722)
- - [SpecDataHolder.sol](https://optimistic.etherscan.io/address/0xEE0c743A3E50133B63eDFcc0006aA331Adf1e4BC)

- Live contracts on **Goerli**
- - [Badges.sol](https://goerli.etherscan.io/address/0xa6773847d3D2c8012C9cF62818b320eE278Ff722)
- - [Raft.sol](https://goerli.etherscan.io/address/0xBb8997048e5F0bFe6C9D6BEe63Ede53BD0236Bb2)
- - [SpecDataHolder.sol](https://goerli.etherscan.io/address/0x7F9279B24D1c36Fa3E517041fdb4E8788dc63D25)

## 2.0.0

- Refactor SpecDataHolder and Badges to use interfaces instead of importing whole contracts
- uncomment constructor in Badges
- improve consistency with variable naming

- Live contracts on **Optimism**
- - [Badges.sol](https://optimistic.etherscan.io/address/0x7F9279B24D1c36Fa3E517041fdb4E8788dc63D25)
- - [Raft.sol](https://optimistic.etherscan.io/address/0xa74caa864a2562999faf38280a3aa3d09c248daa)
- - [SpecDataHolder.sol](https://optimistic.etherscan.io/address/0xEE0c743A3E50133B63eDFcc0006aA331Adf1e4BC)

- Live contracts on **Goerli**
- - [Badges.sol](https://goerli.etherscan.io/address/0xa6773847d3D2c8012C9cF62818b320eE278Ff722)
- - [Raft.sol](https://goerli.etherscan.io/address/0xe620d9CACA4C2B02601C08095a0d5aA14C59270E)
- - [SpecDataHolder.sol](https://goerli.etherscan.io/address/0x7F9279B24D1c36Fa3E517041fdb4E8788dc63D25)

## 1.0.1

- Added `/out/Raft.sol/Raft.json` to repo
- Live contracts on Optimism
- Live contracts on **Optimism**
- - [Badges.sol](https://optimistic.etherscan.io/address/0x639a1703CfdeDaE61A535d53890130b4257f15eb)
- - [Raft.sol](https://optimistic.etherscan.io/address/0xa74caa864A2562999faf38280A3aA3d09c248daA)
- - [SpecDataHolder.sol](https://optimistic.etherscan.io/address/0xdB8346EAF8C4A7eF82B17Ce7843dF8A9d00dC524)

## 1.0.0

- Initial release on Optimism
- Contracts upgradeable using [UUPS pattern](https://docs.openzeppelin.com/contracts/4.x/api/proxy#UUPSUpgradeable)
- Add unit tests with Forge
- Fix security vulnerabilites
- Live contracts on **Optimism**
- - [Badges.sol](https://optimistic.etherscan.io/address/0x639a1703CfdeDaE61A535d53890130b4257f15eb)
- - [Raft.sol](https://optimistic.etherscan.io/address/0xa74caa864a2562999faf38280a3aa3d09c248daa)
- - [SpecDataHolder.sol](https://optimistic.etherscan.io/address/0xdB8346EAF8C4A7eF82B17Ce7843dF8A9d00dC524)
- Live contracts on **Goerli**
- - [Badges.sol](https://goerli.etherscan.io/address/0x835bD6b20206417ff9168B174cE67D812D746dc5)
- - [Raft.sol](https://goerli.etherscan.io/address/0xe620d9CACA4C2B02601C08095a0d5aA14C59270E)
- - [SpecDataHolder.sol](https://goerli.etherscan.io/address/0x147e0dF40fdD1340C604726c670329c08176F208)

## 0.6.0

- Make contracts upgradeable
- Seprate spec-related data to SpecDataHolder
- Move logic of ERC-4973 into Badges.sol (ERC-4973 is no longer a dependency)
- Move deployments over to Goerli
- Badges Goerli deployment: https://goerli.etherscan.io/address/0xde30567ebA075D622da01D0836aFcc4356dB9dEC
- Raft Goerli deployment: https://goerli.etherscan.io/address/0xA32Ef0ED6B60dD406f37f31D40044AD8F6530bbe
- SpecDataHolder Goerli deployment: https://goerli.etherscan.io/address/0x562AD9882B50AB12C445c5b0e30acBE02c09b7F9

## 0.5.0

- Add Raft token
- Add Rinkeby deployment/verification scripts for Badges and Raft token
- Improve integration tests
- Badges Rinkeby deployment: https://rinkeby.etherscan.io/address/0x9323497dc6f24df13fcd09d71bb17efa47b659e3
- Rinkeby deployment: https://rinkeby.etherscan.io/address/0x19020014ef77c5dce4fbcf97c2e3d6e67a616fc6

## 0.4.0

- Introduced Hardhat support
- Using Solidity 0.8.15
- Using Yarn instead of npm
- Rinkeby deployment: https://rinkeby.etherscan.io/address/0x19020014ef77c5dce4fbcf97c2e3d6e67a616fc6

## 0.3.0

- Replacing `Chain-Claim` with ` mintWithPermission`` from `ERC4973Permit`
- `BadgeVendorFactory` and `BadgeVendor` are no longer existing in this version

## 0.2.0

- Added `BadgeVendorFactory` and deployed to Rinkeby
- (breaking) In `BadgeVendor` constructor use simple primitive types instead of
  struct.

## 0.1.1

Publishing now to @otterspace-xyz org.

## 0.1.0

First release that makes contracts available on npm. It allows, e.g. importing
the contract ABI as follows:

```nodejs

tasks/proposeUpgrade.ts
import { HardhatRuntimeEnvironment } from 'hardhat/types/runtime'

require('@openzeppelin/hardhat-upgrades')
require('dotenv').config()

const addresses: { [key: string]: string } = {
  Badges: '0xa6773847d3D2c8012C9cF62818b320eE278Ff722',
  SpecDataHolder: '0x7F9279B24D1c36Fa3E517041fdb4E8788dc63D25',
  Raft: '0xBb8997048e5F0bFe6C9D6BEe63Ede53BD0236Bb2',
};

export default async function proposeUpgrade(params: any, hre: HardhatRuntimeEnvironment): Promise<void> {
    // await hre.run('compile')
  const contractName = process.argv[3]
    console.log('process.argv[3]', process.argv[3])
    // hre.upgrades.forceImport(addresses[contractName], 'Badges',)
  const Implementation = await hre.ethers.getContractFactory(contractName)
  const proposal = await hre.defender.proposeUpgrade(
    addresses[contractName],
    Implementation, {
    multisig: process.env.MULTISIG_GOERLI,
      title: `Upgrade ${contractName} implementation to ${Implementation}`,
    }
  )
  console.log('🚀 ~ proposeUpgrade ~ proposal', proposal)
  console.log('Upgrade proposal created at:', proposal.url)
}
